Find mismatched users

kaapstorm opened the pull request:
## Technical Summary

Opening as a draft PR for visibility only. This PR will not be merged.

This is a management command for a particular project. It finds mobile workers who have been allocated to the wrong locations.

It outputs these mobile workers in a format similar to (but not the same as) a user import spreadsheet.

(A user import spreadsheet has columns headed "location_code 1, location_code 2, ..." but in the domain where this is relevant, location site codes are not unique, so this script will use headings "location_id 1, location_id 2, ..." instead.)

kaapstorm committed:
Find mismatched users

kaapstorm commented on corehq/apps/users/management/commands/nphcda_find_mismatches.py:
```diff
@@ -0,0 +1,151 @@
+            # Skip blank rows
+            if not any(v for v in row.values()):
+                continue
+            # Replace blank values with their previous value
```
Sometimes a value is implied, not entered, if it is the same as the value on last populated row.

mkangia replied:
oh wow!

kaapstorm commented on corehq/apps/users/management/commands/nphcda_find_mismatches.py:
```diff
@@ -0,0 +1,151 @@
+
+
+def get_location(domain: str, name: str, parent: Location) -> Location:
+    code = f'{parent.code}|{name.lower()}'
```
`code` is based on appending `name.lower()` to the `name.lower()` values of its parents because in this particular domain site codes cannot be relied on to be unique. (i.e. Duplicate site codes have been found.)

kaapstorm commented on corehq/apps/users/management/commands/nphcda_find_mismatches.py:
```diff
@@ -0,0 +1,151 @@
+Location = namedtuple('Location', 'id name code')
+country = Location('abc123', 'Nambia', 'nambia')
+location_cache = {
+    'nambia|foo|bar|baz': 'def456',
```
Uses simple dictionary cache to reduce impact on the SQLLocation table. Some values can be prepopulated when location names in the file specified by the `users_csv` command line parameter do not match SQLLocation names.

kaapstorm committed:
Use site_code for user corrections

kaapstorm committed:
Multiple location codes per user

kaapstorm committed:
Use a character that does not appear on a keyboard

kaapstorm committed:
User import uses CommCareUser.raw_username

kaapstorm committed:
Return all locations when any are mismatched

mkangia commented on corehq/apps/users/management/commands/nphcda_find_mismatches.py:
```diff
@@ -0,0 +1,197 @@
+                k: v if v else last_full_row[k]
+                for k, v in row.items()
+            }
+            # Fix abbreviated Username value
```
:open_mouth:

kaapstorm committed:
Allow a lot more location assignments

mkangia reviewed (commented):
It seems you are working with an unusual input file here.
Do you want this command to be reviewed or to be reused? if yes, then would be good to share the specifics about the input file. Without them its not possible to understand the decision making here.

kaapstorm committed:
Some tests

This is to prep for the next refactor: This script, or another that uses
some of this code, will need to update case ownership based on location
changes.

Also, to reiterate, this code is not going to be merged.

kaapstorm commented:
> an unusual input file

Yes! The input file is created by a third-party. Its format is roughly like this:

```csv
State, LGA, Ward, Settlement, Username
FOO,   BAR, BAZ,  QUX,        FO/BAZ001
FOO,   BAR, BAZ,  QUUX,       FO/BAZ001
```

This input file is compiled from different sources, and can have some anomalies. This script tries to handle those anomalies.

> Do you want this command to be reviewed ...

Mostly, I'm just sharing so that the team has visibility into what I'm doing. But I would also like to know if anyone notices something that I ought to do differently, 

> ... or to be reused?

I think the script is quite specific to the location structure and case types of one project. Once the project is complete, I don't think this will be reused. (That is why I've used type hints, so that my IDE can help me with object attributes, function parameters, and refactoring.)

kaapstorm committed:
Get location ID map

Map mismatched locations to correct locations. We will need this for
mapping case owners later. Drop CSV output.

kaapstorm committed:
Improve mapping & output

kaapstorm committed:
Output as JSON or Excel

JSON in that format might be problemmatic to read. YAML might be better.

ajeety4 commented on corehq/apps/users/management/commands/nphcda_find_mismatches.py:
```diff
@@ -16,11 +16,11 @@
 )
 location_cache: dict[str, Location] = {
-    'nambia|foo|bar|baz': Location(
+    'nambia·foo·bar·baz': Location(
```
just curious for the reason behind this ?

kaapstorm replied:
It turns out that some location names have been entered with unusual characters, like "<" and "{", so I switched to using a delimiter that isn't available on a keyboard.

ajeety4 commented on custom/nphcda/tests/test_find_mismatches.py:
```diff
@@ -0,0 +1,145 @@
+            user_rows = list(iter_user_rows(csv_filename))
+        assert user_rows == [
+            UserRow(state='FOO', lga='BAR', ward='BAZ', settlement='QUX', username='FO/BAZ001'),
+            UserRow(state='FOO', lga='BAR', ward='BAZ', settlement='QUUX', username='FO/BAZ001'),
```
So the abbreviated username is the correct one ?

Also is it possible for input to have the abbreviated username appear later?
```
State,LGA,Ward,Settlement,Username
FOO,BAR,BAZ,QUUX,BAZ001
FOO,BAR,BAZ,QUX,FO/BAZ001
```

kaapstorm replied:
> So the abbreviated username is the correct one ?

I'm not sure which one you mean by "abbreviated". A mobile worker's full username looks like `fo/baz001@DOMAIN.commcarehq.org` and their raw_username looks like `fo/baz001`. (It has a forward slash in it, which I haven't seen before.)

> is it possible for input to have the abbreviated username appear later?

(So far?) at least one row with `fo/baz001` always appears before any rows with `baz001`. But if the script can't find a CommCareUser with that username, it will fail immediately -- it doesn't catch the UserError exception.

ajeety4 replied:
> I'm not sure which one you mean by "abbreviated".

My bad. I meant to say the `prefix` with the forward slash in it.

Thanks, this answers my query.

ajeety4 reviewed (commented):
Appreciate the test cases. Took a first look and looks good overall. Left a minor suggestion and a query.

kaapstorm committed:
Use YAML, add empty row between users

kaapstorm committed:
Cache SQLLocations

kaapstorm committed:
Make spreadsheet more readable

kaapstorm committed:
Make `has_changes()` a method

kaapstorm committed:
Split location columns

kaapstorm committed:
File format is XLS, not XLSX

kaapstorm committed:
nphcda_move_users_cases management command

kaapstorm committed:
Document expected column names

kaapstorm committed:
Drop empty rows

kaapstorm committed:
Username on every row

kaapstorm committed:
Support lowercase usernames

kaapstorm committed:
Unmapped locations on separate rows

kaapstorm committed:
Accommodate users assigned to the state

kaapstorm committed:
Try to select location with proper site code

kaapstorm committed:
Add Shabara Fada to known codes

kaapstorm committed:
Add "choose_settlement" household case property

kaapstorm committed:
Tests and fixes
